Search and set `aria-current` for links in all stamped partials

warpech opened the pull request:
Search and set `aria-current` for links in all stamped partials (https://github.com/Starcounter/Blending/issues/85)

Searching for the best way to highlight a current menu item in a Starcounter app, I was shifting through the following ideas:

1. declare and use a `<highlight-current-link>` custom element that sets `aria-current` explicitly in declarative shadow DOM
2. put `<highlight-current-link>` custom element to StarcounterClientFiles and use it explicitly in declarative shadow DOM
3. set `aria-current` implicitly in `<palindrom-client>`... this failed because patch-applied event triggers before HTML view is loaded
4. set `aria-current` implicitly in `<starcounter-include>`... I don't like this because s-i is already overloaded with features. Also it is ineffective, because we have many s-i on the page, but this callback should be run only once
5. set `aria-current` implicitly in an event handler in Starcounter app shell

This commit is an impletentation of idea #5

If this PR is accepted, it will be possible to remove the following lines from BlendingEditor and other apps that want to highlight the current link:

- https://github.com/Starcounter/Blending/blob/6fbf01fb6f1dbf97c2c46bdf7fe3218ac342d440/src/BlendingEditor/wwwroot/BlendingEditor/views/MasterPage.html#L27-L79
- https://github.com/Starcounter/Blending/blob/6fbf01fb6f1dbf97c2c46bdf7fe3218ac342d440/src/BlendingEditor/wwwroot/BlendingEditor/views/MasterPage.html#L147

I know that we have agreed (in https://github.com/Starcounter/RebelsLounge/issues/246) to do `<highlight-current-link>` as a standalone Juicy custom element, but my an afterthought is that it does not make sense in this case. Such custom element would be resource intensive and very limited in use to only Starcounter apps.

warpech committed:
search and set `aria-current` for links in all stamped partials (https://github.com/Starcounter/Blending/issues/85)

Searching for the best way to highlight a current menu item in a Starcounter app, I was shifting through the following ideas:

1. declare and use a <highlight-current-link> custom element that sets `aria-current` explicitly in declarative shadow DOM
2. put <highlight-current-link> custom element to StarcounterClientFiles and use it explicitly in declarative shadow DOM
3. set `aria-current` implicitly in <palindrom-client>... this failed because patch-applied event triggers before HTML view is loaded
4. set `aria-current` implicitly in <starcounter-inclide>... I don't like this because s-i is already overloaded with features. Also it is ineffective, because we have many s-i on the page, but this callback should be run only once
5. set `aria-current` implicitly in an event handler in Starcounter app shell

This commit is an impletentation of idea #5

alshakero commented on src/StarcounterClientFiles/wwwroot/sys/starcounter.html:
```diff
@@ -1,2 +1,33 @@
+            if (node.href === currentHref) {
+                nextCurrent = node;
+            }
+        });
```
Wouldn't be faster and cleaner to hand this loop and filter to the browser istead of doing it in the JS level?

```js
lastCurrent = document.querySelector("a[aria-current]")
nextCurrent = document.querySelector(`[href="${currentHref}"]`);
```

warpech replied:
Your script will only find the first instance, ~~mine will find all instances~~  (Edit: actually mine also has the flaw of finding only the last occurrence). I think it is important to find all.

Regarding `[href="${currentHref}"]`, I think this makes the strong comparison by the attribute string and will not match all possible variants of a href, such as:

```html
<a href="/myapp/mypage">Absolute link</a>
<a href="../myapp/mypage">Relative link</a>
<a href="//host:8080/myapp/mypage">Host link</a>
<a href="https:///host:8080/myapp/mypage">Protocol link</a>
```

alshakero commented on src/StarcounterClientFiles/wwwroot/sys/starcounter.html:
```diff
@@ -1,2 +1,33 @@
+
+    window.addEventListener("stamped", (event) => {
+        if (event.target.localName === "starcounter-include") {
+            updateCurrentLinks(window.location.href);
```
Since its a global variable I would put it as a default parameter in the function and remove it from here. 

```js
function updateCurrentLinks(currentHref = window.location.href) {
```

warpech replied:
I believe that the approach that I made is better for testability. It is better for a function to not depend on globals.

alshakero reviewed (commented):
I thought about the approach, and the you went for seems the best solution on hand. I just left a few suggestions.

alshakero committed:
Mark all matching anchors as aria-current

As opposed to marking only the first or last match of
[href="currentURL"]

alshakero commented:
I disabled `<highlight-current-link>` in Blending and tested my suggestion. Works neatly.

warpech commented on src/StarcounterClientFiles/wwwroot/sys/starcounter.html:
```diff
@@ -1,33 +1,24 @@
+            if (event.target.localName === "starcounter-include") {
+                // debouce for 50ms to avoid doing the same thing for every stamped partial
+                clearTimeout(timeout);
+                timeout = setTimeout(updateCurrentLinks, 50, window.location.href)
```
I didn't know you can pass parameters this way! Neat

warpech reviewed (commented):
Lgtm!

Can you upgrade libraries that for SCF 3.2.0 release into this PR?

alshakero committed:
Update bower dependancies:

1. Palindrom to 5.1.0
2. starcounter-include to 5.0.0
3. underwears.css to 0.0.7

warpech committed:
ensure that the paths to /sys/polymer are created

because otherwise the script does not work on a clean "bower install"

warpech committed:
clean bower install

upgraded Underwear.css to properly built 0.0.7
upgraded Uniform.css to 0.0.8
deleted translate-shadowdom

warpech committed:
bower list >! bower-list.txt

warpech committed:
3.2.0
